pkg/manifests/config.go

```diff
@@ -31,23 +31,28 @@ type Config struct {
 	RemoteWrite bool                  `json:"-"`
 	Platform    configv1.PlatformType `json:"-"`
 
-	PrometheusOperatorConfig             *PrometheusOperatorConfig `json:"prometheusOperator"`
-	PrometheusOperatorUserWorkloadConfig *PrometheusOperatorConfig `json:"prometheusOperatorUserWorkload"`
-
-	PrometheusK8sConfig          *PrometheusK8sConfig `json:"prometheusK8s"`
-	PrometheusUserWorkloadConfig *PrometheusK8sConfig `json:"prometheusUserWorkload"`
+	ClusterMonitoringConfiguration *ClusterMonitoringConfiguration `json:"-"`
+	UserWorkloadConfiguration      *UserWorkloadConfiguration      `json:"-"`
+}
 
-	AlertmanagerMainConfig *AlertmanagerMainConfig      `json:"alertmanagerMain"`
-	KubeStateMetricsConfig *KubeStateMetricsConfig      `json:"kubeStateMetrics"`
-	OpenShiftMetricsConfig *OpenShiftStateMetricsConfig `json:"openshiftStateMetrics"`
-	GrafanaConfig          *GrafanaConfig               `json:"grafana"`
-	EtcdConfig             *EtcdConfig                  `json:"-"`
-	HTTPConfig             *HTTPConfig                  `json:"http"`
-	TelemeterClientConfig  *TelemeterClientConfig       `json:"telemeterClient"`
-	K8sPrometheusAdapter   *K8sPrometheusAdapter        `json:"k8sPrometheusAdapter"`
-	UserWorkloadConfig     *UserWorkloadConfig          `json:"techPreviewUserWorkload"`
-	ThanosRulerConfig      *ThanosRulerConfig           `json:"thanosRuler"`
-	ThanosQuerierConfig    *ThanosQuerierConfig         `json:"thanosQuerier"`
+type ClusterMonitoringConfiguration struct {
+	PrometheusOperatorConfig *PrometheusOperatorConfig    `json:"prometheusOperator"`
+	PrometheusK8sConfig      *PrometheusK8sConfig         `json:"prometheusK8s"`
+	AlertmanagerMainConfig   *AlertmanagerMainConfig      `json:"alertmanagerMain"`
+	KubeStateMetricsConfig   *KubeStateMetricsConfig      `json:"kubeStateMetrics"`
+	OpenShiftMetricsConfig   *OpenShiftStateMetricsConfig `json:"openshiftStateMetrics"`
+	GrafanaConfig            *GrafanaConfig               `json:"grafana"`
+	EtcdConfig               *EtcdConfig                  `json:"-"`
+	HTTPConfig               *HTTPConfig                  `json:"http"`
+	TelemeterClientConfig    *TelemeterClientConfig       `json:"telemeterClient"`
+	K8sPrometheusAdapter     *K8sPrometheusAdapter        `json:"k8sPrometheusAdapter"`
+	ThanosQuerierConfig      *ThanosQuerierConfig         `json:"thanosQuerier"`
+	UserWorkloadEnabled      *bool                        `json:"enableUserWorkload"`
+	// TODO: Remove in 4.7 release.
+	PrometheusUserWorkloadConfig         *PrometheusK8sConfig      `json:"prometheusUserWorkload"`
```

Comment:
Instead of duplicating `PrometheusUserWorkloadConfig`, `PrometheusOperatorUserWorkloadConfig` and `ThanosRulerConfig` here and in `UserWorkloadConfiguration` maybe it would be simpler to use only `UserWorkloadConfiguration`? IOW handle the "legacy" fields only when loading the config map but don't expose them here.

Reply:
The point was to be able to easily remove the current logic, hence the TODO comments, as we will stop supporting this in 4.7. And to separate out the two clear tenants we have right now cluster and user workload.

Reply:
An alternative approach would be to remove the user workload monitoring fields from the `ClusterMonitoringConfiguration` struct and only use `Config.UserWorkloadConfiguration`.  `NewConfig()` would try to load the uwm fields until we decide that it shouldn't.

Something like this: https://github.com/simonpasquier/cluster-monitoring-operator/commit/44bd3ddbed3f2ee07ef99d239e15f1b599ccb581

The side effect is that uwm configuration will be taken either from the "legacy" fields or from the new configmap in `openshift-user-workload-monitoring` but it won't be a merge of both. In the current state, `PrometheusOperatorConfig` can be defined in the `openshift-montoring` configmap and `PrometheusConfig` in the `openshift-user-workload-monitoring` configmap, it can be confusing or desired, depending on how you see things :)

(IOW: in other words)

Reply:
hmm .. especially with configuration I much prefer duplication structs. This is a lesson learned the hard way in Kubernetes. Only because things happen to be identical at this point in time, doesn't mean they may not very well develop in different direction in the future, and they almost always do unless the concept is literally the same (eg. TLS configuration in Prometheus). In this case we even know it will be deleted in the future, so I actually prefer @lilic 's currently proposed way.